src/zarr/storage/_obstore.py

```diff
@@ -0,0 +1,419 @@
+    else:
+        resp = await obs.get_async(store, request["path"], options={"range": request["range"]})
+    buffer = await resp.bytes_async()
+    return [
+        {
+            "original_request_index": request["original_request_index"],
+            "buffer": prototype.buffer.from_bytes(buffer),  # type: ignore[arg-type]
+        }
+    ]
+
+
+async def _get_partial_values(
+    store: _UpstreamObjectStore,
+    prototype: BufferPrototype,
+    key_ranges: Iterable[tuple[str, ByteRequest | None]],
+) -> list[Buffer | None]:
+    """Make multiple range requests.
+
+    ObjectStore has a `get_ranges` method that will additionally merge nearby ranges,
+    but it's _per_ file. So we need to split these key_ranges into **per-file** key
+    ranges, and then reassemble the results in the original order.
+
+    We separate into different requests:
+
+    - One call to `obstore.get_ranges_async` **per target file**
+    - One call to `obstore.get_async` for each other request.
+    """
+    key_ranges = list(key_ranges)
+    per_file_bounded_requests: dict[str, list[_BoundedRequest]] = defaultdict(list)
+    other_requests: list[_OtherRequest] = []
+
+    for idx, (path, byte_range) in enumerate(key_ranges):
+        if byte_range is None:
+            other_requests.append(
+                {
+                    "original_request_index": idx,
+                    "path": path,
+                    "range": None,
+                }
+            )
+        elif isinstance(byte_range, RangeByteRequest):
+            per_file_bounded_requests[path].append(
+                {"original_request_index": idx, "start": byte_range.start, "end": byte_range.end}
+            )
+        elif isinstance(byte_range, OffsetByteRequest):
+            other_requests.append(
+                {
+                    "original_request_index": idx,
+                    "path": path,
+                    "range": {"offset": byte_range.offset},
+                }
+            )
+        elif isinstance(byte_range, SuffixByteRequest):
+            other_requests.append(
+                {
+                    "original_request_index": idx,
+                    "path": path,
+                    "range": {"suffix": byte_range.suffix},
+                }
+            )
+        else:
+            raise ValueError(f"Unsupported range input: {byte_range}")
+
+    futs: list[Coroutine[Any, Any, list[_Response]]] = []
+    for path, bounded_ranges in per_file_bounded_requests.items():
+        futs.append(_make_bounded_requests(store, path, bounded_ranges, prototype))
+
+    for request in other_requests:
+        futs.append(_make_other_request(store, request, prototype))  # noqa: PERF401
+
+    buffers: list[Buffer | None] = [None] * len(key_ranges)
+
+    # TODO: this gather a list of list of Response; not sure if there's a way to
+    # unpack these lists inside of an `asyncio.gather`?
+    for responses in await asyncio.gather(*futs):
```

Comment:
I thought we have a `gather` implementation inside zarr that respects a concurrency limit. But now I'm not seeing it (just `zarr.core.common.concurrent_map`, which I think isn't quite right).

Anyway, maybe we'd want to do acquire an `asyncio.Semaphore` with a limit of `zarr.config.get("async.concurrency")` in each of these coroutines.

Reply:
I don't think we have this utility but are instead creating semaphores manually: https://github.com/zarr-developers/zarr-python/blob/f95c453f3835da7f8097a1b5302ea1b7a3d6e57f/src/zarr/core/group.py#L1384

Reply:
I think I fixed this in https://github.com/zarr-developers/zarr-python/pull/1661/commits/c45c7d1bc27a720185dbcf0cdada58e8c84b7f0a if you can check my work.

Reply:
Looks good.